pkg/api/fulcioca_signing_cert.go

```diff
@@ -0,0 +1,129 @@
+	"github.com/spf13/viper"
+
+	"github.com/sigstore/fulcio/pkg/ca/fulcioca"
+	"github.com/sigstore/fulcio/pkg/ctl"
+	"github.com/sigstore/fulcio/pkg/generated/restapi/operations"
+	"github.com/sigstore/fulcio/pkg/log"
+	"github.com/sigstore/fulcio/pkg/oauthflow"
+	"github.com/sigstore/fulcio/pkg/pkcs11"
+)
+
+func FulcioCASigningCertHandler(params operations.SigningCertParams, principal *oidc.IDToken) middleware.Responder {
+
+	// none of the following cases should happen if the authentication path is working correctly; checking to be defensive
+	if principal == nil {
+		return handleFulcioAPIError(params, http.StatusInternalServerError, errors.New("no principal supplied to request"), invalidCredentials)
+	}
+
+	emailAddress, emailVerified, err := oauthflow.EmailFromIDToken(principal)
+	if !emailVerified {
+		return handleFulcioAPIError(params, http.StatusBadRequest, errors.New("email_verified claim was false"), emailNotVerified)
+	} else if err != nil {
+		return handleFulcioAPIError(params, http.StatusBadRequest, err, invalidCredentials)
+	}
+
+	publicKey := *params.CertificateRequest.PublicKey
+	pkixPubKey, err := x509.ParsePKIXPublicKey(*publicKey.Content)
+	if err != nil {
+		return handleFulcioAPIError(params, http.StatusBadRequest, err, malformedPublicKey)
+	}
+
+	// Perform a proof challenge verification
+	if err := fulcioca.CheckSignature(pkixPubKey, *params.CertificateRequest.SignedEmailAddress, emailAddress); err != nil {
+		return handleFulcioAPIError(params, http.StatusBadRequest, err, invalidSignature)
+	}
+
+	// Use a generic Cert failure message as we don't want to tell the client about a HSM failure and risk exposure of
+	// internal security controls
+	p11Ctx, err := pkcs11.InitHSMCtx()
+	if err != nil {
+		return handleFulcioAPIError(params, http.StatusInternalServerError, err, failedToCreateCert)
+	}
+	defer p11Ctx.Close()
+
+	rootID := []byte(viper.GetString("hsm-caroot-id"))
+
+	// get the existing root CA from the HSM
+	rootCA, err := p11Ctx.FindCertificate(rootID, nil, nil)
+	if err != nil {
+		return handleFulcioAPIError(params, http.StatusInternalServerError, err, failedToCreateCert)
+	}
+
+	// get the private key object from HSM
+	privKey, err := p11Ctx.FindKeyPair(nil, []byte("FulcioCA"))
+	if err != nil {
+		return handleFulcioAPIError(params, http.StatusInternalServerError, err, failedToCreateCert)
+	}
+
+	// Generate the client signing certificate
+
+	clientCert, err := fulcioca.CreateClientCertificate(rootCA, emailAddress, pkixPubKey, privKey)
+	if err != nil {
+		return handleFulcioAPIError(params, http.StatusInternalServerError, err, failedToCreateCert)
+	}
+
+	// Format in PEM
+	certPEM := pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: clientCert,
+	})
+
+	// Format in PEM
+	rootPEM := pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: rootCA.Raw,
+	})
+
+	// Submit to CTL
+	log.Logger.Info("Submitting CTL inclusion for OIDC grant: ", subject)
```

Comment:
please change all the logging statements to be:
`log.RequestIDLogger(params.HTTPRequest).Info()` so we can correlate with the HTTP transaction

Reply:
done